lib/engine-jit/src/artifact.rs

```diff
@@ -325,15 +326,34 @@ impl Artifact for JITArtifact {
         &self.func_data_registry
     }
     fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        // let mut s = flexbuffers::FlexbufferSerializer::new();
-        // self.serializable.serialize(&mut s).map_err(|e| SerializeError::Generic(format!("{:?}", e)));
-        // Ok(s.take_buffer())
-        let bytes = bincode::serialize(&self.serializable)
-            .map_err(|e| SerializeError::Generic(format!("{:?}", e)))?;
-
         // Prepend the header.
         let mut serialized = Self::MAGIC_HEADER.to_vec();
-        serialized.extend(bytes);
+
+        serialized.resize(32, 0);
+        let mut writable_leb = &mut serialized[16..];
+        let serialized_data = self.serializable.serialize()?;
+        let length = serialized_data.len();
+        leb128::write::unsigned(&mut writable_leb, length as u64).expect("Should write number");
+
+        let align = std::mem::align_of::<SerializableModule>() as u64;
+
+        let offset = pad_and_extend(&mut serialized, &serialized_data, align);
+        assert_eq!(offset, 32);
+
         Ok(serialized)
     }
 }
+
+/// It pads the data with the desired alignment
+pub fn pad_and_extend(prev_data: &mut Vec<u8>, data: &[u8], align: u64) -> u64 {
+    // We assert that align is a power of 2
+    debug_assert_eq!(align & (align - 1), 0);
```

Comment:
Luckily Rust has you covered [`u64::is_power_of_two`](https://doc.rust-lang.org/std/primitive.u64.html#method.is_power_of_two):

⛳ 
```Rust
debug_assert!(align.is_power_of_two());
```

Alternatively, if you know that `std::mem::align_of` always returns a power of two, rather than taking `align` as a runtime parameter, you can take the type as a generic.

So you'd update the function to look like:

```Rust
pub fn pad_and_extend<T>(prev_data: &mut Vec<u8>, data: &[u8]) -> u64 {
   let align = std::mem::align_of::<T>() as u64;
   ...
}
```

Which would then prevent the function from ever being called incorrectly in this way. (you could obviously pass the wrong type and get the wrong align but 🤷 )

You may then need to change the caller to explicitly fill the generic parameter, so like:
```Rust
let offset = pad_and_extend::<SerializableModule>(&mut serialized, &serialized_data);
```

Reply:
Love this!

Reply:
Fixed!